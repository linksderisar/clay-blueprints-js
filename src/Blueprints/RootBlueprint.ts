import _ from 'lodash';
import AbstractBlueprint from "./Abstracts/AbstractBlueprint";

export default class RootBlueprint extends AbstractBlueprint {
    private _store: object;
    private _meta: object[];
    private _head: object;
    private _componentTree: AbstractBlueprint;

    constructor(componentTree) {
        super();
        this._store = {};
        this._meta = [];
        this._head = {};
        this._componentTree = componentTree;

    }

    addMetas(metas) {
        this._meta = [ ...this._meta, ...metas ];
    }

    get store() {
        return this._store;
    }

    set store(value) {
        this._store = value;
    }

    get meta() {
        return this._meta;
    }

    set meta(value) {
        this._meta = value;
    }

    get head() {
        return this._head;
    }

    set head(value) {
        this._head = value;
    }

    get componentTree() {
        return this._componentTree;
    }

    set componentTree(value) {
        this._componentTree = value;
    }

    toObject(): any {
        return {
            'store': this._store,
            'componentTree': _.isEmpty(this._componentTree) ? {} : this._componentTree.toObject(),
            'meta': this._meta,
            'head': this._head,
        }
    }
}
