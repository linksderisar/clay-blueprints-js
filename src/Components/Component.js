import _ from 'lodash';
import ComponentBlueprint from '../Blueprints/ComponentBlueprint';

export default class Component {
    constructor(type) {
        this._children = {};
        this._scopedSlots = {};

        this._bindable = {
            'prop': 'addBindProp',
            'props': 'addBindProps',
            'attribute': 'addBindAttribute',
            'attributes': 'addBindAttributes'
        };

        this._blueprint = new ComponentBlueprint(this._type);
        this._type = type;
    }

    get type() {
        return this._type;
    }

    set type(value) {
        this._type = value;
        return this;
    }

    get children() {
        return this._children;
    }

    set children(value) {
        this._children = value;
        return this;
    }

    get scopedSlots() {
        return this._scopedSlots;
    }

    set scopedSlots(value) {
        this._scopedSlots = value;
        return this;
    }

    get bindable() {
        return this._bindable;
    }

    set bindable(value) {
        this._bindable = value;
        return this;
    }

    get blueprint() {
        return this._blueprint;
    }

    set blueprint(value) {
        this._blueprint = value;
        return this;
    }

    prop(key, value) {
        this.blueprint.addProps({[key]: value});
        return this;
    }

    props(newProps) {
        this.blueprint.addProps(newProps);
        return this;
    }

    attribute(key, value) {
        this.blueprint.addAttributes({[key]: value});
        return this;
    }

    attributes(props) {
        this.blueprint.addAttributes(props);
        return this;
    }

    // name clash with getter
    extendChildren(newChildren) {
        this.children = newChildren;

        this.blueprint.children = _.map(this.children, (value) => value.blueprint);
        return this;
    }

    classes(classes) {
        this.blueprint.classes = classes;
        return this;
    }

    style(cssAttribute, value) {
        this.blueprint.style = {[cssAttribute]: value};
        return this;
    }

    styles(styles) {
        this.blueprint.styles = styles;
        return this;
    }

    ref(r) {
        this.blueprint.ref = r;
        return this;
    }

    key(k) {
        this.blueprint.key = k;
        return this;
    }

    refInfor(r = true) {
        this.blueprint.refInfor = r;
        return this;
    }

    affect(a) {
        this.blueprint.affect = a;
        return this;
    }

    text(t) {
        this.blueprint.text = t;
        return this;
    }

    toObject() {
        return this.blueprint.toObject();
    }

    toJson() {
        return this.blueprint.toJson();
    }
}
