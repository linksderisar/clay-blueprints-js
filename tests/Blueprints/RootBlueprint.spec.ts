import RootBlueprint from '../../src/Blueprints/RootBlueprint';
import Component from '../../src/Components/Component';

let blueprint = null;

beforeEach(() => blueprint = new RootBlueprint());

it('initializes an empty root blueprint', () => {
    expect(blueprint.toJson()).toMatchSnapshot();
});

it('sets the store', () => {
    blueprint.store = {'data': 'value'};
    expect(blueprint.toJson()).toMatchSnapshot();
});

it('sets meta', () => {
    blueprint.meta = {'meta': 'meta-value'};
    expect(blueprint.toJson()).toMatchSnapshot();
});

it('sets head', () => {
    blueprint.head = {'title': 'here is a title'};
    expect(blueprint.toJson()).toMatchSnapshot();
});

it('sets the component tree', () => {
    blueprint.componentTree = new Component('div');
    expect(blueprint.toObject()).toMatchSnapshot({
        componentTree: {
            id: expect.any(String)
        }
    });
});

it('sets the component tree during construct', () => {
    blueprint = new RootBlueprint(new Component('div'));
    expect(blueprint.toObject()).toMatchSnapshot({
        componentTree: {
            id: expect.any(String)
        }
    });
});
