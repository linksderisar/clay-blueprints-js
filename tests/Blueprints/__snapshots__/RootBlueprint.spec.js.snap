// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initializes an empty root blueprint 1`] = `"{\\"store\\":{},\\"componentTree\\":{},\\"meta\\":{},\\"head\\":{}}"`;

exports[`sets head 1`] = `"{\\"store\\":{},\\"componentTree\\":{},\\"meta\\":{},\\"head\\":{\\"title\\":\\"here is a title\\"}}"`;

exports[`sets meta 1`] = `"{\\"store\\":{},\\"componentTree\\":{},\\"meta\\":{\\"meta\\":\\"meta-value\\"},\\"head\\":{}}"`;

exports[`sets the component tree 1`] = `
Object {
  "componentTree": Object {
    "id": Any<String>,
    "type": "",
  },
  "head": Object {},
  "meta": Object {},
  "store": Object {},
}
`;

exports[`sets the component tree during construct 1`] = `
Object {
  "componentTree": Object {
    "id": Any<String>,
    "type": "",
  },
  "head": Object {},
  "meta": Object {},
  "store": Object {},
}
`;

exports[`sets the store 1`] = `"{\\"store\\":{\\"data\\":\\"value\\"},\\"componentTree\\":{},\\"meta\\":{},\\"head\\":{}}"`;
